### DISCUSSION TITLE
Expected behaviour when X-B3-Sampled is not set

### DISCUSSION THREAD

**Discussion by paulgrav** [open-telemetry/opentelemetry-java-instrumentation#3867] [Q&A]
As the title suggests should I expect traces to be sampled when only `X-B3-TraceId` and `X-B3-SpanId` are set?

We have some runtime environment constraints (long story, don’t ask) and we are still running SDK/agent 0.8.0. When `X-B3-Sampled` is set traces are exported, when that header is missing traces are not exported. Is there a way to force them to be exported besides setting `X-B3-Sampled`?

**Comment by mateuszrzeszutek** [comment]
Here's what the [b3 spec](https://github.com/openzipkin/b3-propagation#sampling-state-1) says about the `X-B3-Sampled` header:

> An accept sampling decision is encoded as X-B3-Sampled: 1 and a deny as X-B3-Sampled: 0. Absent means defer the decision to the receiver of this header.

It seems OTel SDK has decided to drop those spans that don't have that header.

> Is there a way to force them to be exported besides setting `X-B3-Sampled`?

You can set the `otel.traces.sampler` property to `always_on` - you'll get all the spans though.

**Reply by jkwatson** [reply]
This might be a bug in the B3 Propagator. Here is the code : https://github.com/open-telemetry/opentelemetry-java/blob/main/extensions/trace-propagators/src/main/java/io/opentelemetry/extension/trace/propagation/Common.java#L36-L39

This code is being used by both the B3 Propagator and the W3C TraceContext propagator. The flag isn't nullable coming from W3C TraceContext, but it is from B3, so we probably need to explicitly handle the null `sampled` value coming from B3. Should null -> sampled, then, to meet the requirements from B3?

**Reply by paulgrav** [reply]
Right. I don’t think `otel.traces.sampler` was available in the 0.8.0 version.

**Reply by mateuszrzeszutek** [reply]
> Right. I don’t think `otel.traces.sampler` was available in the 0.8.0 version.

Oh, right, that's a pretty old version. Hmm, if you were using just SDK you would be able to set the always-on sampler yourself, but with the agent it might be near impossible to do so.

**Reply by jkwatson** [reply]
https://github.com/open-telemetry/opentelemetry-java/issues/3517 . I do think this is a deeper issue, though. Our propagation APIs don't have a way to handle this "you decide on the sampling" case from an upstream trace.
